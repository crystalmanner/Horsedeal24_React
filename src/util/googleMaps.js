import { types as sdkTypes } from '../util/sdkLoader';

const { LatLng: SDKLatLng, LatLngBounds: SDKLatLngBounds } = sdkTypes;

const placeOrigin = place => {
  if (place && place.geometry && place.geometry.location) {
    return new SDKLatLng(place.geometry.location.lat(), place.geometry.location.lng());
  }
  return null;
};

const placeBounds = place => {
  if (place && place.geometry && place.geometry.viewport) {
    const ne = place.geometry.viewport.getNorthEast();
    const sw = place.geometry.viewport.getSouthWest();
    return new SDKLatLngBounds(
      new SDKLatLng(ne.lat(), ne.lng()),
      new SDKLatLng(sw.lat(), sw.lng())
    );
  }
  return null;
};

/**
 * Get a detailed place object
 *
 * @param {String} placeId - ID for a place received from the
 * autocomplete service
 * @param {String} sessionToken - token to tie different autocomplete character searches together
 * with getPlaceDetails call
 *
 * @return {Promise<util.propTypes.place>} Promise that
 * resolves to the detailed place, rejects if the request failed
 */
export const getPlaceDetails = (placeId, sessionToken) =>
  new Promise((resolve, reject) => {
    const serviceStatus = window.google.maps.places.PlacesServiceStatus;
    const el = document.createElement('div');
    const service = new window.google.maps.places.PlacesService(el);
    const fields = ['address_component', 'formatted_address', 'geometry', 'place_id'];
    const sessionTokenMaybe = sessionToken ? { sessionToken } : {};

    service.getDetails({ placeId, fields, ...sessionTokenMaybe }, (place, status) => {
      if (status !== serviceStatus.OK) {
        reject(
          new Error(`Could not get details for place id "${placeId}", error status was "${status}"`)
        );
      } else {
        resolve({
          address: place.formatted_address,
          origin: placeOrigin(place),
          bounds: placeBounds(place),
        });
      }
    });
  });

const predictionSuccessful = status => {
  const { OK, ZERO_RESULTS } = window.google.maps.places.PlacesServiceStatus;
  return status === OK || status === ZERO_RESULTS;
};

/**
 * Get place predictions for the given search
 *
 * @param {String} search - place name or address to search
 * @param {String} sessionToken - token to tie different autocomplete character searches together
 * with getPlaceDetails call
 * @param {Object} searchConfigurations - defines the search configurations that can be used with
 * the autocomplete service. Used to restrict search to specific country (or countries).
 *
 * @return {Promise<{ search, predictions[] }>} - Promise of an object
 * with the original search query and an array of
 * `google.maps.places.AutocompletePrediction` objects
 */
export const getPlacePredictions = (search, sessionToken, searchConfigurations) =>
  new Promise((resolve, reject) => {
    const service = new window.google.maps.places.AutocompleteService();
    const sessionTokenMaybe = sessionToken ? { sessionToken } : {};
    const searchConfigurations = {
      // types: ['(cities)'],
      componentRestrictions: {country: 'ch'},
    };
    service.getPlacePredictions(
      { input: search, ...sessionTokenMaybe, ...searchConfigurations },
      (predictions, status) => {
        if (!predictionSuccessful(status)) {
          reject(new Error(`Prediction service status not OK: ${status}`));
        } else {
          const results = {
            search,
            predictions: predictions || [],
          };
          resolve(results);
        }
      }
    );
  });

/**
 * Deprecation: use function from src/util/maps.js
 * Cut some precision from bounds coordinates to tackle subtle map movements
 * when map is moved manually
 *
 * @param {LatLngBounds} sdkBounds - bounds to be changed to fixed precision
 * @param {Number} fixedPrecision - integer to be used on tofixed() change.
 *
 * @return {SDKLatLngBounds} - bounds cut to given fixed precision
 */
export const sdkBoundsToFixedCoordinates = (sdkBounds, fixedPrecision) => {
  const fixed = n => Number.parseFloat(n.toFixed(fixedPrecision));
  const ne = new SDKLatLng(fixed(sdkBounds.ne.lat), fixed(sdkBounds.ne.lng));
  const sw = new SDKLatLng(fixed(sdkBounds.sw.lat), fixed(sdkBounds.sw.lng));

  return new SDKLatLngBounds(ne, sw);
};

/**
 * Deprecation: use function from src/util/maps.js
 * Check if given bounds object have the same coordinates
 *
 * @param {LatLngBounds} sdkBounds1 - bounds #1 to be compared
 * @param {LatLngBounds} sdkBounds2 - bounds #2 to be compared
 *
 * @return {boolean} - true if bounds are the same
 */
export const hasSameSDKBounds = (sdkBounds1, sdkBounds2) => {
  if (!(sdkBounds1 instanceof SDKLatLngBounds) || !(sdkBounds2 instanceof SDKLatLngBounds)) {
    return false;
  }
  return (
    sdkBounds1.ne.lat === sdkBounds2.ne.lat &&
    sdkBounds1.ne.lng === sdkBounds2.ne.lng &&
    sdkBounds1.sw.lat === sdkBounds2.sw.lat &&
    sdkBounds1.sw.lng === sdkBounds2.sw.lng
  );
};

/**
 * Calculate a bounding box in the given location
 *
 * @param {latlng} center - center of the bounding box
 * @param {distance} distance - distance in meters from the center to
 * the sides of the bounding box
 *
 * @return {LatLngBounds} bounding box around the given location
 *
 */
export const locationBounds = (latlng, distance) => {
  const bounds = new window.google.maps.Circle({
    center: new window.google.maps.LatLng(latlng.lat, latlng.lng),
    radius: distance,
  }).getBounds();

  const ne = bounds.getNorthEast();
  const sw = bounds.getSouthWest();

  return new SDKLatLngBounds(new SDKLatLng(ne.lat(), ne.lng()), new SDKLatLng(sw.lat(), sw.lng()));
};
