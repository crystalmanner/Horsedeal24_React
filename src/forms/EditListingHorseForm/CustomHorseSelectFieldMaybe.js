import React from 'react';
import { required } from '../../util/validators';
import { FieldSelect } from '../../components';

import css from './EditListingHorseForm.css';

const CustomHorseSelectFieldMaybe = props => {
  const { name, id, values, selectLabel, selectPlaceholder, selectRequired } = props;

  return values ? (
    <FieldSelect
      className={css.selectField}
      name={name}
      id={id}
      label={selectLabel}
      validate={required(selectRequired)}
    >
      <option disabled value="">
        {selectPlaceholder}
      </option>
      {values.map(c => (
        <option key={c.key} value={c.key}>
          {c.label}
        </option>
      ))}
    </FieldSelect>
  ) : null;
};

export default CustomHorseSelectFieldMaybe;
