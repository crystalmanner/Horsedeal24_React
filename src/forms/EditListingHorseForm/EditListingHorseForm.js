import React from 'react';
import { arrayOf, bool, func, shape, string } from 'prop-types';
import { compose } from 'redux';
import { Form as FinalForm } from 'react-final-form';
import { intlShape, injectIntl, FormattedMessage } from '../../util/reactIntl';
import classNames from 'classnames';
import { propTypes } from '../../util/types';
import { maxLength, required, composeValidators } from '../../util/validators';
import { Form, Button, FieldTextInput } from '../../components';
import CustomCategorySelectFieldMaybe from './CustomHorseSelectFieldMaybe';

import css from './EditListingHorseForm.css';

const TITLE_MAX_LENGTH = 60;

const EditListingHorseFormComponent = props => (
  <FinalForm
    {...props}
    render={fieldRenderProps => {
      const {
        className,
        disabled,
        handleSubmit,
        intl,
        invalid,
        pristine,
        saveActionMsg,
        updated,
        updateInProgress,
        fetchErrors,
        genders,
        ages,
        breeds,
        hights,
        colors,
      } = fieldRenderProps;

      const titleMessage = intl.formatMessage({ id: 'EditListingHorseForm.title' });
      const titlePlaceholderMessage = intl.formatMessage({
        id: 'EditListingHorseForm.titlePlaceholder',
      });
      const titleRequiredMessage = intl.formatMessage({
        id: 'EditListingHorseForm.titleRequired',
      });

      const genderMessage = intl.formatMessage({ id: 'EditListingHorseForm.gender' });
      const genderPlaceholderMessage = intl.formatMessage({
        id: 'EditListingHorseForm.genderPlaceholder',
      });
      const genderRequiredMessage = intl.formatMessage({
        id: 'EditListingHorseForm.genderRequired',
      });

      const ageMessage = intl.formatMessage({ id: 'EditListingHorseForm.age' });
      const agePlaceholderMessage = intl.formatMessage({
        id: 'EditListingHorseForm.agePlaceholder',
      });
      const ageRequiredMessage = intl.formatMessage({
        id: 'EditListingHorseForm.ageRequired',
      });

      const breedMessage = intl.formatMessage({ id: 'EditListingHorseForm.breed' });
      const breedPlaceholderMessage = intl.formatMessage({
        id: 'EditListingHorseForm.breedPlaceholder',
      });
      const breedRequiredMessage = intl.formatMessage({
        id: 'EditListingHorseForm.breedRequired',
      });

      const hightMessage = intl.formatMessage({ id: 'EditListingHorseForm.hight' });
      const hightPlaceholderMessage = intl.formatMessage({
        id: 'EditListingHorseForm.hightPlaceholder',
      });
      const hightRequiredMessage = intl.formatMessage({
        id: 'EditListingHorseForm.hightRequired',
      });

      const colorMessage = intl.formatMessage({ id: 'EditListingHorseForm.color' });
      const colorPlaceholderMessage = intl.formatMessage({
        id: 'EditListingHorseForm.colorPlaceholder',
      });
      const colorRequiredMessage = intl.formatMessage({
        id: 'EditListingHorseForm.colorRequired',
      });

      const maxLengthMessage = intl.formatMessage(
        { id: 'EditListingHorseForm.maxLength' },
        {
          maxLength: TITLE_MAX_LENGTH,
        }
      );

      const maxLength60Message = maxLength(maxLengthMessage, TITLE_MAX_LENGTH);

      const { updateListingError, createListingDraftError, showListingsError } = fetchErrors || {};
      const errorMessageUpdateListing = updateListingError ? (
        <p className={css.error}>
          <FormattedMessage id="EditListingHorseForm.updateFailed" />
        </p>
      ) : null;

      // This error happens only on first tab (of EditListingWizard)
      const errorMessageCreateListingDraft = createListingDraftError ? (
        <p className={css.error}>
          <FormattedMessage id="EditListingHorseForm.createListingDraftError" />
        </p>
      ) : null;

      const errorMessageShowListing = showListingsError ? (
        <p className={css.error}>
          <FormattedMessage id="EditListingHorseForm.showListingFailed" />
        </p>
      ) : null;

      const classes = classNames(css.root, className);
      const submitReady = updated && pristine;
      const submitInProgress = updateInProgress;
      const submitDisabled = invalid || disabled || submitInProgress;

      return (
        <Form className={classes} onSubmit={handleSubmit}>
          {errorMessageCreateListingDraft}
          {errorMessageUpdateListing}
          {errorMessageShowListing}
          <FieldTextInput
            id="title"
            name="title"
            className={css.title}
            type="text"
            label={titleMessage}
            placeholder={titlePlaceholderMessage}
            maxLength={TITLE_MAX_LENGTH}
            validate={composeValidators(required(titleRequiredMessage), maxLength60Message)}
            autoFocus
          />
          <CustomCategorySelectFieldMaybe
            id="gender"
            name="gender"
            values={genders}
            selectLabel={genderMessage}
            selectPlaceholder={genderPlaceholderMessage}
            selectRequired={genderRequiredMessage}
          />
          <CustomCategorySelectFieldMaybe
            id="age"
            name="age"
            values={ages}
            selectLabel={ageMessage}
            selectPlaceholder={agePlaceholderMessage}
            selectRequired={ageRequiredMessage}
          />
          <CustomCategorySelectFieldMaybe
            id="breed"
            name="breed"
            values={breeds}
            selectLabel={breedMessage}
            selectPlaceholder={breedPlaceholderMessage}
            selectRequired={breedRequiredMessage}
          />
          <CustomCategorySelectFieldMaybe
            id="hight"
            name="hight"
            values={hights}
            selectLabel={hightMessage}
            selectPlaceholder={hightPlaceholderMessage}
            selectRequired={hightRequiredMessage}
          />
          <CustomCategorySelectFieldMaybe
            id="color"
            name="color"
            values={colors}
            selectLabel={colorMessage}
            selectPlaceholder={colorPlaceholderMessage}
            selectRequired={colorRequiredMessage}
          />
          <Button
            className={css.submitButton}
            type="submit"
            inProgress={submitInProgress}
            disabled={submitDisabled}
            ready={submitReady}
          >
            {saveActionMsg}
          </Button>
        </Form>
      );
    }}
  />
);

EditListingHorseFormComponent.defaultProps = { className: null, fetchErrors: null };

EditListingHorseFormComponent.propTypes = {
  className: string,
  intl: intlShape.isRequired,
  onSubmit: func.isRequired,
  saveActionMsg: string.isRequired,
  updated: bool.isRequired,
  updateInProgress: bool.isRequired,
  fetchErrors: shape({
    createListingDraftError: propTypes.error,
    showListingsError: propTypes.error,
    updateListingError: propTypes.error,
  }),
  genders: arrayOf(
    shape({
      key: string.isRequired,
      label: string.isRequired,
    })
  ),
  ages: arrayOf(
    shape({
      key: string.isRequired,
      label: string.isRequired,
    })
  ),
  breeds: arrayOf(
    shape({
      key: string.isRequired,
      label: string.isRequired,
    })
  ),
  hights: arrayOf(
    shape({
      key: string.isRequired,
      label: string.isRequired,
    })
  ),
  colors: arrayOf(
    shape({
      key: string.isRequired,
      label: string.isRequired,
    })
  ),
};

export default compose(injectIntl)(EditListingHorseFormComponent);
