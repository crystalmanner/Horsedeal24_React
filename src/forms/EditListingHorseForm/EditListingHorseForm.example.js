import EditListingHorseForm from './EditListingHorseForm';

export const Empty = {
  component: EditListingHorseForm,
  props: {
    onSubmit: values => {
      console.log('Submit EditListingHorseForm with (unformatted) values:', values);
    },
    saveActionMsg: 'Save description',
    updated: false,
    updateInProgress: false,
  },
  group: 'forms',
};
