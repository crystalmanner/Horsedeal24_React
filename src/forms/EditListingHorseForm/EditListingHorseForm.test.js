import React from 'react';
import { renderDeep } from '../../util/test-helpers';
import { fakeIntl } from '../../util/test-data';
import EditListingHorseForm from './EditListingHorseForm';

const noop = () => null;

describe('EditListingHorseForm', () => {
  it('matches snapshot', () => {
    const tree = renderDeep(
      <EditListingHorseForm
        intl={fakeIntl}
        dispatch={noop}
        onSubmit={v => v}
        saveActionMsg="Save description"
        updated={false}
        updateInProgress={false}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
