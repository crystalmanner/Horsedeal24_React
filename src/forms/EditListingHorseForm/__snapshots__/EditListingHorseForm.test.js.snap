// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditListingHorseForm matches snapshot 1`] = `
<form
  action="/"
  className=""
  method="post"
  onSubmit={[Function]}
>
  <div
    className=""
  >
    <label
      htmlFor="title"
    >
      EditListingHorseForm.title
    </label>
    <input
      autoFocus={true}
      className=""
      id="title"
      maxLength={60}
      name="title"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="EditListingHorseForm.titlePlaceholder"
      type="text"
      value=""
    />
  </div>
  <button
    className=""
    disabled={true}
    type="submit"
  >
    Save description
  </button>
</form>
`;
